### PR TITLE
feat: expose querykey getter and fix some querykey stuff

### LINKED ISSUE #3295: feat: expose queryKey
### Describe the feature you'd like to request

not all methods available on the `queryClient` are available on `useContext()` from trpc. While it might be desirable to add more methods to it (see: https://github.com/trpc/trpc/issues/3012), it doesn't have to be done for _all_ methods - only those that profit from type-safety.

There are a couple of methods exposed from TanStack/query that do not require additional typesafety, but would need access to the `queryKey`. For example:

- `queryClient.setQueryDefaults(key, defaults)`
- `queryClient.isFetching({ queryKey })`
  - or even `useIsFetching({ queryKey })`

additional type-safety is not needed because for the first example, `defaults` work the same for all keys, and for the second example, only a _number_ is returned.

It would be good if we could use these methods with queries created from trpc.

### Describe the solution you'd like to see

One way to achieve this is to expose the queryKey that was created internally, for example:

```ts
const queryClient = useQueryClient()
const utils = trpc.useContext()

const isFetching = useIsFetching(utils.post.byId.getQueryKey()) // QueryKey => [["post","byId"]]
queryClient.setQueryDefaults(utils.post.getQueryKey(), { staleTime: Infinity }) // QueryKey => [["post"]]
```

### Describe alternate solutions

we could expose _all_ methods via `trpc.useContext()`, but I think it's an uphill battle and not really needed. Some things (but not all of them) will be covered with:
- https://github.com/trpc/trpc/issues/3012

### Additional information

reference:
- https://github.com/trpc/trpc/issues/3012#issuecomment-1319653598

### 👨‍👧‍👦 Contributing

- [ ] 🙋‍♂️ Yes, I'd be down to file a PR implementing this feature!

### LINKED ISSUE #3455: bug: query invalidation doesn't accept filters
### Provide environment information

```
  System:
    OS: macOS 12.5.1
    CPU: (8) arm64 Apple M1 Pro
    Memory: 77.77 MB / 32.00 GB
    Shell: 5.8.1 - /bin/zsh
  Binaries:
    Node: 16.13.0 - ~/.nvm/versions/node/v16.13.0/bin/node
    Yarn: 1.22.19 - ~/.nvm/versions/node/v16.13.0/bin/yarn
    npm: 8.1.0 - ~/.nvm/versions/node/v16.13.0/bin/npm
  Browsers:
    Brave Browser: 108.1.46.144
    Chrome: 108.0.5359.124
    Firefox Developer Edition: 108.0
    Safari: 15.6.1
  npmPackages:
    @tanstack/react-query: 4.20.4 => 4.20.4 
    @trpc/client: 10.6.0 => 10.6.0 
    @trpc/react-query: 10.6.0 => 10.6.0 
    @trpc/server: 10.6.0 => 10.6.0 
    next: 12.3.0 => 12.3.0 
    react: 17.0.2 => 17.0.2 
```

### Describe the bug

passing filters to `invalidate` doesn't seem to have an effect - no invalidation is happening

examples:

```
utils.invalidate({ predicate: () => true })
utils.invalidate({ refetchType: 'all' })
```

### Link to reproduction

https://codesandbox.io/s/fancy-pond-dznks2?file=/src/pages/index.tsx

### To reproduce

click the `invalidate` button.

## expected:

the query should be refetched

## actually:

no refetch happens

### Additional information

it works without filters, so this is fine:

```
utils.invalidate()
```

### 👨‍👧‍👦 Contributing

- [ ] 🙋‍♂️ Yes, I'd be down to file a PR fixing this bug!

### LINKED ISSUE #3456: bug: query invalidation can only target trpc queries
### Provide environment information

```
  System:
    OS: macOS 12.5.1
    CPU: (8) arm64 Apple M1 Pro
    Memory: 77.77 MB / 32.00 GB
    Shell: 5.8.1 - /bin/zsh
  Binaries:
    Node: 16.13.0 - ~/.nvm/versions/node/v16.13.0/bin/node
    Yarn: 1.22.19 - ~/.nvm/versions/node/v16.13.0/bin/yarn
    npm: 8.1.0 - ~/.nvm/versions/node/v16.13.0/bin/npm
  Browsers:
    Brave Browser: 108.1.46.144
    Chrome: 108.0.5359.124
    Firefox Developer Edition: 108.0
    Safari: 15.6.1
  npmPackages:
    @tanstack/react-query: 4.20.4 => 4.20.4 
    @trpc/client: 10.6.0 => 10.6.0 
    @trpc/react-query: 10.6.0 => 10.6.0 
    @trpc/server: 10.6.0 => 10.6.0 
    next: 12.3.0 => 12.3.0 
    react: 17.0.2 => 17.0.2 
```

### Describe the bug

given that you have an environment where you're mixing `useQuery` from `react-query` and queries from `trpc`, `utils.invalidate()` _used to_ target all queries, but now, it only targets the queries created by trpc.

### Link to reproduction

https://codesandbox.io/s/festive-newton-rodx1p?file=/src/pages/index.tsx

### To reproduce

- click the `invalidate` button
- observe that only the query created by trpc is invalidated

### Additional information

This used to work with `10.0.0-rc.3`

from discord:

```
quick question: I think the change of query keys (that we now have { type: 'query'} etc. at the end) kind of introduced an issue when using trpc together with "raw" react-query queries. For reference, we are gradually upgrading to tprc and still have some "untyped" useQuery calls lying around.

Our issue is that we call utils.invalidate() to invalidate all queries in some cases. It used to target all trpc and react-query queries, but now, it only targets the trpc queries. I've debugged into the queryClient and what we get into the invalidation as a key is [[],{}] and this does match "all trpc queries", but not "all queries".

Our workaround is to use the real queryClient for invalidation now, just wanted to let you know about that - maybe it just worked coincidentally before
```

### 👨‍👧‍👦 Contributing

- [ ] 🙋‍♂️ Yes, I'd be down to file a PR fixing this bug!

### PR DESCRIPTION
**Pull request by juliusmarminge**
- Closes #3295
- Closes #3455
- Closes #3456

## 🎯 Changes

What changes are made in this PR? Is it a feature or a bug fix?

- Adds a `getQueryKey` helper for routers and procedures

- fixes some other queryKey related bugs (see #3455 and #3456)

Docs link: https://www-git-julius-querykeys-trpc.vercel.app/docs/useContext#-the-function-i-want-isnt-here

### API

Since we can't distinguish routers and procedures in runtime, we can't dynamically decide the QueryType to call `getArrayQueryKey` with. Therefore, I made these explicit arguments that the user inputs:

```ts
getQueryKey(undefined, "any")   // for routers
getQueryKey(TInput, "infinite") // for infinite queries
getQueryKey(TInput, "query")    // for regular queries
```

The default QueryType is `any`.

Usage is as follows:

```ts
import { useQueryClient } from "@tanstack/react-query";
import { trpc } from "~/utils/trpc";

const MyComp = () => {
  const qc = useQueryClient();
  const qKey = trpc.post.getQueryKey(undefined, "any");
  setQueryDefaults(qKey, { staleTime: 30 * 60 * 1000 });
}
```

Another usage is for route prefetching:

https://github.com/trpc/trpc/blob/377e271b97d9392263595b44b2868795e75e818e/examples/tanstack-router/client/src/utils/router.tsx#L55-L61

## ✅ Checklist

- [x] I have followed the steps listed in the [Contributing guide](https://github.com/trpc/trpc/blob/main/CONTRIBUTING.md).
- [x] If necessary, I have added documentation related to the changes made.
- [x] I have added or updated the tests related to the changes made.

### CODE REVIEW

File: packages/tests/server/react/useContext.test.tsx
Comment (juliusmarminge) [inline packages/tests/server/react/useContext.test.tsx:914]: @TkDodo This should pass, right?

Comment (TkDodo) [reply packages/tests/server/react/useContext.test.tsx:914]: no, because you don't have a subscription. `qc.isFetching` just reads the value once. You'd want:

```
const isFetching = useIsFetching({ queryKey: key });
```

you can import it directly:

```
import { useIsFetching } from '@tanstack/react-query'
```

Comment (juliusmarminge) [reply packages/tests/server/react/useContext.test.tsx:914]: oh right.. thanks for clarifying 🙏🏼

File: packages/react-query/src/shared/proxy/utilsProxy.ts
Comment (KATT) [inline packages/react-query/src/shared/proxy/utilsProxy.ts:174]: On my phone so I might be missing something, but wouldn't it make sense to have both arguments optional?

Comment (KATT) [reply packages/react-query/src/shared/proxy/utilsProxy.ts:174]: Cc @TkDodo wdyt?

Comment (juliusmarminge) [reply packages/react-query/src/shared/proxy/utilsProxy.ts:174]: Why would we default to? "query"?

The input would be optional if the procedure's input is optional, but we need the type (I wrote a bit about this up top or in discord with @sachinraja)

Comment (KATT) [reply packages/react-query/src/shared/proxy/utilsProxy.ts:174]: If it's getting the key for sale of invalidation or resetting, `any` is a good default

If its for getting or changing the cache `any` might not be the best

Comment (TkDodo) [reply packages/react-query/src/shared/proxy/utilsProxy.ts:174]: what is `"any"` going to return? does it match both `query` and `infinite` ? If so I think it would be a good default for `getQueryKey`

Comment (JonParton) [reply packages/react-query/src/shared/proxy/utilsProxy.ts:174]: `any` returns the query key without the type object in the key, so as long as fuzzy matching is enabled for the methods then it will match both!

Do all methods match fuzzy? Or for the "query" versions only match exact? This may need some documentation!

Comment (TkDodo) [reply packages/react-query/src/shared/proxy/utilsProxy.ts:174]: > Do all methods match fuzzy?

all methods _can_, but it's not the default everywhere. Generally speaking:
- everything that has a plural in its name is fuzzy, like `getQueriesData`, `invalidateQueries` or `setQueriesDefault`
- everything that is singular defaults to exact, like `getQueryData`

**Comment by JonParton** [comment]
> ### Alternatives
> Since this method doesn't use the query client, it could go on the regular trpc object:
> 
> ```ts
> const qKey = trpc.post.getQueryKey(...);
> ```
> 
> but I prefer it being on the utils one, feels like it "belongs" there? (personal preference maybe?).
> 
> It does require an additional hook to be called though:
> 
> ```ts
> // this
> import { useQueryClient } from "@tanstack/react-query";
> import { trpc } from "~/utils/trpc";
> const MyComp = () => {
>   const qc = useQueryClient();
>   const utils = trpc.useContext();
>   const qKey = utils.post.getQueryKey(undefined, "any");
>   setQueryDefaults(qKey, { staleTime: 30 * 60 * 1000 });
> }
> 
> // over this
> import { useQueryClient } from "@tanstack/react-query";
> import { trpc } from "~/utils/trpc";
> const MyComp = () => {
>   const qc = useQueryClient();
>   const qKey = trpc.post.getQueryKey(undefined, "any");
>   setQueryDefaults(qKey, { staleTime: 30 * 60 * 1000 });
> }
> ```
> 
> What do others think?

First @juliusmarminge; awesome job!! 

On the above, I also see the point around it living on the `useContext()` object mainly because we tend to call the object returned `utils` in our examples. 

However, I think there is value in it being directly on the trpc object or as a seperate helper entirely (that could be instaitated with AppRouter) . The reason being that this then opens it up to use outside of a react component, and I have already come across some use cases ( [Example here ](https://github.com/trpc/trpc/issues/3012#issuecomment-1279960638)  ) for doing this. I also know there is always a request around interacting with the query client outside of react context as well in many issues that get posted here... So it would be allowing the wider use of the escape hatch in those cases too!

File: packages/react-query/src/shared/proxy/utilsProxy.ts
Comment (JonParton) [reply packages/react-query/src/shared/proxy/utilsProxy.ts:174]: The fuzzy matching behaviour queried above for `xxxxxQuery` type methods also effects if the `input` param should be optional!... Although I think it should be as I know some deffinitely fuzzy match!

Comment (JonParton) [reply packages/react-query/src/shared/proxy/utilsProxy.ts:174]: Hmmn this may mean making things optional / default to a query key of `any` for the main path I expect this being used for (we provide many of the tanstack methods for individual queries already so this is likely to be used in the fuzzy matching ones!) but then put some good documentation on this noting the gotcha between needing exact keys for methods that are named `xxxxxQuery`

**Comment by juliusmarminge** [comment]
> The reason being that this then opens it up to use outside of a react component, and I have already come across some use cases ( Example here ) for doing this. I also know there have been several requests in issues posted on the repo around interacting with the tanstack query queryClient outside of react context... So this would allow that wider use of the escape hatch too!

That's a fair point. That would allow users to use tRPC with @tanstack/react-router too for data loaders I believe. This is probably the way. What's your thought here @KATT @TkDodo ?

**Comment by sachinraja** [comment]
I would put it on the `trpc` object, I think the data loaders stuff is a strong argument in favor of that.

File: examples/tanstack-router/README.md
Comment (juliusmarminge) [inline examples/tanstack-router/README.md:1]: Just a playground for seeing the usage - I don't think we should keep this but instead contribute to https://github.com/TanStack/router/tree/beta/examples/react/with-trpc-react-query

File: examples/tanstack-router/client/src/utils/router.tsx
Comment (juliusmarminge) [inline examples/tanstack-router/client/src/utils/router.tsx]: here we got the usage in the loaders

File: examples/tanstack-router/client/src/utils/router.tsx
Comment (juliusmarminge) [inline examples/tanstack-router/client/src/utils/router.tsx:64]: this is how i thought it'd be useful

**Review by juliusmarminge — COMMENTED** [review]
added a temporary example - should revert before merge (i think?)

**Review by sachinraja — APPROVED** [review]

File: examples/tanstack-router/client/src/utils/router.tsx
Comment (juliusmarminge) [inline examples/tanstack-router/client/src/utils/router.tsx:1]:
```suggestion
/**
 * @see https://tanstack.com/router/v1/docs/examples/react/basic
 **/
/* eslint-disable react-hooks/rules-of-hooks */
```

**Comment by aulneau** [comment]
this is great! hope it can be merged soon

File: www/docs/reactjs/useContext.mdx
Comment (TkDodo) [inline www/docs/reactjs/useContext.mdx:107]: can we document the default value of `type` here? Looking at the code, I think it defaults to `"any"` ?

Comment (juliusmarminge) [reply www/docs/reactjs/useContext.mdx:107]: Yes definetely should be documented

File: examples/tanstack-router/client/src/utils/router.tsx
Comment (KATT) [inline examples/tanstack-router/client/src/utils/router.tsx:108]: Shouldn't we be able to use `.prefetch` here? Or is it b/c it's only on the context? Is there a way to make that happen?

File: examples/tanstack-router/client/src/utils/router.tsx
Comment (KATT) [inline examples/tanstack-router/client/src/utils/router.tsx:113]: Is this run on the frontend and/or the backend? `import`ing backend stuff on something that is "universal" always scares me

File: examples/tanstack-router/client/src/utils/router.tsx
Comment (KATT) [inline examples/tanstack-router/client/src/utils/router.tsx:9]: Importing stuff from the server here shouldn't be needed, right? See comments further down

File: examples/tanstack-router/README.md
Comment (KATT) [reply examples/tanstack-router/README.md:1]: Seems like they haven't solved the data loading either? https://github.com/TanStack/router/blob/484af7fd844c3d4b02d67846ec268bc7310b9aba/examples/react/with-trpc-react-query/client/main.tsx#L111-L114

Comment (juliusmarminge) [reply examples/tanstack-router/README.md:1]: Maybe just remove the example for now then and revisit later?

File: examples/tanstack-router/client/src/utils/router.tsx
Comment (juliusmarminge) [reply examples/tanstack-router/client/src/utils/router.tsx]: That doesn't work since this is outside the react tree and we don't have our context stuff here. Might be possible to do something similar to the SSG helpers (or maybe they even work?)

**Comment by juliusmarminge** [comment]
I removed the example for now @KATT, maybe we can revisit it later but I dont think it should block this PR.

File: examples/tanstack-router/client/src/utils/router.tsx
Comment (KATT) [reply examples/tanstack-router/client/src/utils/router.tsx]: Hmmmmmm..... I don't like embedding this example when the DX is not optimal

File: www/docs/reactjs/useContext.mdx
Comment (KATT) [inline www/docs/reactjs/useContext.mdx:123]: We should call the var just `queryClient`

**Review by KATT — APPROVED** [review]

**Review by KATT — DISMISSED** [review]

File: pnpm-lock.yaml
Comment (KATT) [inline pnpm-lock.yaml:122]: Bad

Comment (juliusmarminge) [reply pnpm-lock.yaml]: weird..........

fixed now though

**Review by KATT — DISMISSED** [review]

File: packages/tests/server/react/useContext.test.tsx
Comment (juliusmarminge) [inline packages/tests/server/react/useContext.test.tsx:393]: would love a bit nicer test here

File: packages/tests/server/interop/react/invalidateQueries.test.tsx
Comment (juliusmarminge) [inline packages/tests/server/interop/react/invalidateQueries.test.tsx:230]: not sure why it's failing but cba fixing this right now - is it necessary to fix this for interop? I doubt anyone will notice it not working if they haven't already xd

Comment (KATT) [reply packages/tests/server/interop/react/invalidateQueries.test.tsx:230]: I'd like not to have conscious regression but make an issue of it and we can do it later

**Review by KATT — APPROVED** [review]

File: packages/tests/server/interop/react/invalidateQueries.test.tsx
Comment (juliusmarminge) [reply packages/tests/server/interop/react/invalidateQueries.test.tsx:230]: Done, #3460

File: packages/tests/server/interop/react/invalidateQueries.test.tsx
Comment (juliusmarminge) [inline packages/tests/server/interop/react/invalidateQueries.test.tsx:203]: nvm I found it and fixed it here instead

File: packages/tests/server/react/invalidateQueries.test.tsx
Comment (KATT) [inline packages/tests/server/react/invalidateQueries.test.tsx:189]: Isn't this changing the behaviour? 🤔 

Looks like a breaking change?

File: packages/tests/server/interop/react/invalidateQueries.test.tsx
Comment (KATT) [inline packages/tests/server/interop/react/invalidateQueries.test.tsx:195]: Breaking chaange?

**Review by KATT — COMMENTED** [review]
Hmm, is this a breaking change?

File: packages/tests/server/react/invalidateQueries.test.tsx
Comment (juliusmarminge) [reply packages/tests/server/react/invalidateQueries.test.tsx:189]: I mean yea theoretically although it wasnt working before so its more of a fix?

Comment (juliusmarminge) [reply packages/tests/server/react/invalidateQueries.test.tsx:189]: Mentioned this here: https://discord.com/channels/867764511159091230/1055437438266392646/1055450208424706099

We always take the first arg as input no matter what the types said so any filters applied would actually end up being supressed

File: packages/tests/server/interop/react/invalidateQueries.test.tsx
Comment (juliusmarminge) [reply packages/tests/server/interop/react/invalidateQueries.test.tsx:195]: See above

**Review by KATT — APPROVED** [review]

**Comment by TkDodo** [comment]
awesome - thanks for this ❤️

**Pull request merged**
